python/rikai/spark/sql/model.py

```diff
@@ -0,0 +1,260 @@
+        """Validate model spec
+
+        Raises
+        ------
+        SpecError
+            If the spec is not well-formatted.
+        """
+        logger.debug("Validating spec: %s", self._spec)
+        try:
+            validate(instance=self._spec, schema=SPEC_PAYLOAD_SCHEMA)
+        except ValidationError as e:
+            raise SpecError(e.message) from e
+
+    @property
+    def version(self) -> str:
+        """Returns spec version."""
+        return str(self._spec["version"])
+
+    @property
+    def name(self) -> str:
+        """Return model name"""
+        return self._spec["name"]
+
+    @property
+    def model_uri(self) -> str:
+        """Return Model artifact URI"""
+        return self._spec["model"]["uri"]
+
+    @property
+    def model_type(self) -> "ModelType":
+        """Return model type"""
+        mtype = self._spec["model"].get("type", None)
+        if mtype:
+            return parse_model_type(self.flavor, mtype)
+        return AnonymousModelType(
+            self._spec.get("schema", None),
+            self.pre_processing,
+            self.post_processing,
+        )
+
+    @abstractmethod
+    def load_model(self) -> Any:
+        """Load the model artifact specified in this spec"""
+
+    @property
+    def flavor(self) -> str:
+        """Model flavor"""
+        return self._spec["model"].get("flavor", "")
+
+    @property
+    def schema(self) -> str:
+        """Return the output schema of the model."""
+        if "schema" in self._spec:
+            return parse_schema(self._spec["schema"])
+        return self.model_type.schema
+
+    @property
+    def options(self) -> Dict[str, Any]:
+        """Model options"""
+        return self._spec["options"]
+
+    @property
+    def pre_processing(self) -> Optional[Callable]:
+        """Return pre-processing transform if exists"""
+        if (
+            "transforms" not in self._spec
+            or "pre" not in self._spec["transforms"]
+            or self._spec["transforms"]["pre"] is None
+        ):
+            # Passthrough
+            return _identity
+        f = find_func(self._spec["transforms"]["pre"])
+        return f(self.options)
+
+    @property
+    def post_processing(self) -> Optional[Callable]:
+        """Return post-processing transform if exists"""
+        if (
+            "transforms" not in self._spec
+            or "post" not in self._spec["transforms"]
+            or self._spec["transforms"]["post"] is None
+        ):
+            # Passthrough
+            return _identity
+        f = find_func(self._spec["transforms"]["post"])
+        return f(self.options)
+
+
+class ModelType(ABC):
+    @abstractmethod
+    def load_model(self, spec: ModelSpec, **kwargs):
+        pass
+
+    @abstractmethod
+    def schema(self) -> str:
+        pass
+
+    def dataType(self) -> "pyspark.sql.types.DataType":
+        return parse_schema(self.schema())
+
+    @abstractmethod
+    def transform(self) -> Callable:
```

Comment:
it seems transform is `pre_processing` and predict is `post_processing`, could you add some comments here?

Reply:
done